### PR TITLE
Ensure idle-processing (like solution syncing) does not execute while paused due to global operations (like msbuild compiles)

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Potential fix for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1511901

In that issue we are seeing shutdown delayed due to contention caused by roslyn.  One issue seen in the trace was excess CPU and allocations in this time period caused by us processing the results of a build being finished (including syncing the workspace to OOP).  This happens because we turn off all processing while the build is in progress, but then start it up again once teh build finishes.  However, it's still the case that such processing should be not continue because VS is not in an 'idle' state due to all the work happening on shutdown.

To try to alleviate this, i'm merging the 'pausing and idle' logic into one check in one location to prevent work from happening at all if *either* we are paused due to global operations, *or* we have not been idle long enough.

This also included a bunch fo simplification as these systems are highly complex, often for dubious reasons.

### CODE REVIEW

File: src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs [hunk 1]
```diff
@@ -53,8 +53,6 @@ public SolutionChecksumUpdater(Workspace workspace, IGlobalOptionService globalO
             Start();
         }
 
-        private CancellationToken ShutdownCancellationToken => CancellationToken;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs:56]: extremely confusing to have two CTs in scope that are the same thing, but with different names, making it unclear what it means when some code in this type refers to one, and some to the other.

File: src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs [hunk 2]
```diff
@@ -64,19 +62,16 @@ protected override async Task ExecuteAsync()
                 _currentToken = null;
             }
 
-            // wait for global operation to finish
-            await GlobalOperationTask.ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs:68]: intentionally removed (will make more sense later).  we are only called into ExecuteAsync when we're not paused due to a global operation being in flight.

File: src/Features/Core/Portable/SolutionCrawler/GlobalOperationAwareIdleProcessor.cs [hunk 3]
```diff
@@ -15,59 +15,28 @@ internal abstract class GlobalOperationAwareIdleProcessor : IdleProcessor
     {
         private readonly IGlobalOperationNotificationService _globalOperationNotificationService;
 
-        private TaskCompletionSource<object?>? _globalOperation;
-        private Task _globalOperationTask;
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/GlobalOperationAwareIdleProcessor.cs:19]: instead of a separate tracking trask, the logic for pausing moved up into the base class (IdleProcessor).

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 4]
```diff
@@ -39,6 +48,11 @@ internal abstract class IdleProcessor
         protected abstract Task WaitAsync(CancellationToken cancellationToken);
         protected abstract Task ExecuteAsync();
 
+        /// <summary>
+        /// Will be called in a serialized fashion (i.e. never concurrently).
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:52]: note: this is hte stated invariant of hte system from before.  i've done no analysis to ensure it's actually true.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 5]
```diff
@@ -48,24 +62,48 @@ protected void Start()
...
+                if (value)
+                    OnPaused();
+            }
+        }
+
+        protected async Task WaitForIdleAsync(IExpeditableDelaySource expeditableDelaySource)
+        {
+            while (!CancellationToken.IsCancellationRequested)
+            {
+
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:97]:
```suggestion
```

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 6]
```diff
@@ -48,24 +62,48 @@ protected void Start()
...
+                // If we're not paused, and enough time has elapsed, then we're done.  Otherwise, ensure we wait at
+                // least s_minimumDelay and check again in the future.
                 var diff = _timeSinceLastAccess.Elapsed;
-                if (diff >= BackOffTimeSpan)
-                {
+                if (!Paused && diff >= BackOffTimeSpan)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:101]: core difference here.  if we're still paused, we dont' return, even if enough time has passed.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 7]
```diff
@@ -48,24 +62,48 @@ protected void Start()
...
                     return;
-                }
 
-                // TODO: will safestart/unwarp capture cancellation exception?
                 var timeLeft = BackOffTimeSpan - diff;
-                if (!await expeditableDelaySource.Delay(TimeSpan.FromMilliseconds(Math.Max(s_minimumDelay.TotalMilliseconds, timeLeft.TotalMilliseconds)), CancellationToken).ConfigureAwait(false))
+                var delayTimeSpan = TimeSpan.FromMilliseconds(Math.Max(s_minimumDelay.TotalMilliseconds, timeLeft.TotalMilliseconds));
+                if (!await expeditableDelaySource.Delay(delayTimeSpan, CancellationToken).ConfigureAwait(false))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:106]: extracted to local just for readability.  No change in semantics here.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 8]
```diff
@@ -86,11 +124,13 @@ private async Task ProcessAsync()
                 {
                     // wait for next item available
                     await WaitAsync(CancellationToken).ConfigureAwait(false);
+                    CancellationToken.ThrowIfCancellationRequested();
 
                     using (Listener.BeginAsyncOperation("ProcessAsync"))
                     {
                         // we have items but workspace is busy. wait for idle.
                         await WaitForIdleAsync(Listener).ConfigureAwait(false);
+                        CancellationToken.ThrowIfCancellationRequested();
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:133]: added early exits so that we don't even try to start work if we're shutting down.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs [hunk 9]
```diff
@@ -49,6 +49,10 @@ private sealed class HighPriorityProcessor : IdleProcessor
                         Start();
                     }
 
+                    protected override void OnPaused()
+                    {
+                    }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs:54]: the base method is abstract so that i had to go to all types to make it clear in them that they're doing nothing (like before).

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs [hunk 10]
```diff
@@ -71,31 +71,22 @@ protected async Task WaitForHigherPriorityOperationsAsync()
                     {
                         using (Logger.LogBlock(FunctionId.WorkCoordinator_WaitForHigherPriorityOperationsAsync, CancellationToken))
                         {
-                            do
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:74]: no clue why this was a `do/while(true)`.  `while(true)` has the same semantics.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs [hunk 11]
```diff
@@ -71,31 +71,22 @@ protected async Task WaitForHigherPriorityOperationsAsync()
...
                                 // we wait for global operation and higher queue operation if there is anything going on
-                                if (!GlobalOperationTask.IsCompleted || !HigherQueueOperationTask.IsCompleted)
-                                {
-                                    await Task.WhenAll(GlobalOperationTask, HigherQueueOperationTask).ConfigureAwait(false);
-                                }
-
-                                // if there are no more work left for higher queue, then it is our time to go ahead
-                                if (!HigherQueueHasWorkItem)
-                                {
+                                await HigherQueueOperationTask.ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:79]: this method (WaitForHigherPriorityOperationsAsync) is only called at teh start of ExecuteAsync.  So there is no need to do any checking around global operations being in flight.  the top level idle processor loop only calls into this when there's no need to delay work.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs [hunk 12]
```diff
@@ -71,31 +71,22 @@ protected async Task WaitForHigherPriorityOperationsAsync()
...
+
+                                // if there are no more work left for higher queue, and we didn't enter a state where we
+                                // should wait for idle again, then our time to go ahead.
+                                if (!HigherQueueHasWorkItem && !ShouldWaitForIdle())
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:83]: important cahnge.  while waiting for the higher pri work to be done.  we should go back to idling if other stuff happened (like us getting paused for a global operation).

File: src/Workspaces/Core/Portable/Notification/AbstractGlobalOperationNotificationService.cs [hunk 13]
```diff
@@ -1,20 +0,0 @@
...
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-#nullable disable
-
-using System;
-
-namespace Microsoft.CodeAnalysis.Notification
-{
-    internal abstract class AbstractGlobalOperationNotificationService : IGlobalOperationNotificationService
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/AbstractGlobalOperationNotificationService.cs:11]: only one impl of this.  so just got rid of the abstract type.

File: src/Workspaces/Core/Portable/Notification/GlobalOperationEventArgs.cs [hunk 14]
```diff
@@ -1,17 +0,0 @@
...
-// See the LICENSE file in the project root for more information.
-
-using System;
-using System.Collections.Immutable;
-
-namespace Microsoft.CodeAnalysis.Notification
-{
-    internal class GlobalOperationEventArgs : EventArgs
-    {
-        public ImmutableArray<string> Operations { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/GlobalOperationEventArgs.cs:12]: these values were never read.  so i removed htis EventArgs subclass and replaced with EventArgs.

File: src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.GlobalOperationRegistration.cs [hunk 15]
```diff
@@ -6,34 +6,26 @@
...
         {
             _service = service;
-            _done = false;
-            this.Operation = operation;
 
             _source = new CancellationTokenSource();
             _logging = Logger.LogBlock(FunctionId.GlobalOperationRegistration, operation, _source.Token);
         }
 
-        public string Operation { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.GlobalOperationRegistration.cs:29]: never read.

File: src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.GlobalOperationRegistration.cs [hunk 16]
```diff
@@ -6,34 +6,26 @@
...
-
-        /// <summary>
-        /// Used to indicate that the global operation completed fully.  The only effect this has is how this operation
-        /// will be logged when <see cref="Dispose"/> is called. If this has been called, then <see cref="Dispose"/>
-        /// will log that we completed without cancellation.  If this has not been called, then <see cref="Dispose"/>
-        /// will log that we were canceled.
-        /// </summary>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.GlobalOperationRegistration.cs:36]: note: this whole system is suspect.  i doubt we've eer looked at this telemetry.  i'm not removing it though in case it matters at some point.

File: src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.cs [hunk 17]
```diff
@@ -2,80 +2,40 @@
...
-                {
-                    Contract.ThrowIfFalse(_operations.Count == 1);
-                    RaiseGlobalOperationStartedAsync();
-                }
-
-                return registration;
-            }
-        }
-
-        private Task RaiseGlobalOperationStartedAsync()
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.cs:55]: no clue why these synchronous methods were async/task-returning.

File: src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.cs [hunk 18]
```diff
@@ -114,10 +110,8 @@ public override void Done(GlobalOperationRegistration registration)
 
                 if (_registrations.Count == 0)
                 {
-                    var operations = _operations.AsImmutable();
                     _operations.Clear();
-
-                    RaiseGlobalOperationStoppedAsync(operations);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Notification/GlobalOperationNotificationService.cs:120]: nothing was ever looking at these tasks, so i cleaned up.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:79]: note: this system is entirely suspect to me @mavasani .  We have all these processor subclasses, and all of them then coordinate with each other to make sure only one is running (and the highest priority one at that).  It's unclear to me why we don' just have a top level processor that just has three queues (or one priority queue) and it just grabs the appropriate next item out of each queue to run.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 19]
```diff
@@ -48,24 +62,49 @@ protected void Start()
...
+
+        protected async Task WaitForIdleAsync(IExpeditableDelaySource expeditableDelaySource)
+        {
+            while (true)
+            {
+                this.CancellationToken.ThrowIfCancellationRequested();
+
+                // If we're not paused, and enough time has elapsed, then we're done.  Otherwise, ensure we wait at
+                // least s_minimumDelay and check again in the future.
                 var diff = _timeSinceLastAccess.Elapsed;
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:101]: No longer used?
```suggestion
```

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 20]
```diff
@@ -48,24 +62,49 @@ protected void Start()
...
+                    // we were started while in the *middle* of some global operation.
+                    if (value)
+                        Contract.ThrowIfTrue(_paused_doNotAccessDirectly);
+
+                    _paused_doNotAccessDirectly = value;
                 }
 
+                // Let subclasses know we're paused so they can change what they're doing accordingly.
+                if (value)
+                    OnPaused();
```
Comment (mavasani) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:89]: Given that the setter is actually invoking a method here, and also taking a lock above to set the field, I think it makes more sense to remove the setter and convert this into a separate method.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs (hunk 11 above)
Comment (mavasani) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:79]: I believe this has been there since the original design. I would be fine if we consolidate these into single processor. One thing that I do notice is that `HighPriorityProcessor` is not a sub-type of `GlobalOperationAwareIdleProcessor`, while the normal and low priority processors are global operation aware. They also seem to have very different implementations, I am not sure if any of these variations were introduced due to any performance investigations. That would be my biggest concern around a big refactoring here. Otherwise, I completely agree that this is a super complicated system, and would really love if it can be simplified, or removed in future.

**Review by mavasani — COMMENTED** [review]
So, overall this change and simplification looks fine to me. However, I am not able to follow what is the core change that fixes the bug cited in the PR description. Can you point to the specific code change in the PR which is the bug fix?

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 20 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:89]: Sounds good.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs [hunk 21]
```diff
@@ -48,24 +62,49 @@ protected void Start()
...
+            while (true)
+            {
+                this.CancellationToken.ThrowIfCancellationRequested();
+
+                // If we're not paused, and enough time has elapsed, then we're done.  Otherwise, ensure we wait at
+                // least s_minimumDelay and check again in the future.
                 var diff = _timeSinceLastAccess.Elapsed;
-                if (diff >= BackOffTimeSpan)
-                {
+                if (!ShouldWaitForIdle())
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:102]: @mavasani This is the core part of the change.  Instead of just saying "hey, enough time has passed, start doing your processing again" now we say "hey, even if enough time has passed, ensure we don't start doing work if we're also paused".

And the reverse is also true.  Previously, if we were paused and became unpaused, we would start up work, even if not enough time had passed since hte "last access time" had elapsed.

Basically, we had two separate checks, in two different places, and they were not coordinated.  Now, it's centralized.  In order to run solution crawler work, we need to be both non-paused *and* have seen enough time pass.

File: src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.AbstractPriorityProcessor.cs:79]: yeah, i don't have a good plan here yet.  It's just... extremely complex and confusing.

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 21 above)
Comment (mavasani) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs:102]: Thanks! Sounds good to me.

**Review by mavasani — APPROVED** [review]

File: src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs (hunk 20 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs]: done.
